tests/test_cli.py

```diff
@@ -0,0 +1,213 @@
+GENERAL_INVALID_ARGS = [
+    "foo",
+    "0",
+]
+
+# Output markers:
+HELP_SIGNATURE = (
+    "usage: browser-history " "[-h] [-t TYPE] [-b BROWSER] [-f FORMAT] [-o OUTPUT]"
+)
+HISTORY_TITLE = "history:\n"
+BOOKMARKS_TITLE = "bookmarks:\n"
+HISTORY_HEADER = "Timestamp,URL"
+BOOKMARKS_HEADER = "Timestamp,URL,Title,Folder"
+
+CSV_HISTORY_HEADER = HISTORY_TITLE + HISTORY_HEADER
+CSV_BOOKMARKS_HEADER = BOOKMARKS_TITLE + BOOKMARKS_HEADER
+
+
+def test_no_argument():
+    """Test the root command gives basic output."""
+    output = subprocess.check_output([CMD_ROOT])
+    assert CSV_HISTORY_HEADER in output.decode("utf-8")
+
+
+def test_type_argument():
+    """Test arguments for the type option."""
+    for type_opt in VALID_CMD_OPTS[1]:
+        for type_arg in VALID_TYPE_ARGS:
+            output = subprocess.check_output([CMD_ROOT, type_opt, type_arg]).decode(
+                "utf-8"
+            )
+            if type_arg == "history":  # in ('history', 'all')
+                assert output.startswith(HISTORY_TITLE)
+            if type_arg == "bookmarks":  # in ('bookmarks', 'all')
+                assert output.startswith(BOOKMARKS_TITLE)
+
+
+@pytest.mark.parametrize("browser_arg", VALID_BROWSER_ARGS)
+def test_browser_argument(browser_arg):
+    """Test arguments for the browser option."""
+    for browser_opt in VALID_CMD_OPTS[2]:
+        output = subprocess.Popen(
+            [CMD_ROOT, browser_opt, browser_arg],
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
+        )
+        stdout, stderr = output.communicate()
+        if output.returncode == 0:
+            assert CSV_HISTORY_HEADER in stdout.decode("utf-8")
+        elif any(
+            browser_unavailable_err in stderr.decode("utf-8")
+            for browser_unavailable_err in (
+                "browser is not supported",
+                "browser is not installed",
+            )
+        ):
+            # In case the tester does not have access to the browser
+            # in question, which makes the command fail, but in a way
+            # that is expected and gives a recognised error message.
+            pytest.skip(
+                "Unable to test against this browser because it is "
+                "not available locally"
+            )
+        else:  # command fails for another reason i.e. a test failure
+            pytest.fail(
+                "The browser is available but the command to fetch "
+                "the history from it has failed."
+            )
+
+
+def test_format_argument():
+    """Tests arguments for the format option."""
+    # First check format of default:
+    csv_output = subprocess.check_output([CMD_ROOT]).decode("utf-8")
+    # Sniffer determines format, less intensive than reading in csv.reader
+    # and we don't mind the CSV dialect, so just check call doesn't error
+    read_csv = csv.Sniffer()
+    read_csv.sniff(csv_output, delimiters=",")
+    read_csv.has_header(csv_output)
```

Comment:
Does `has_header` do an assert internally? Or does it through an error when there's no header?

Reply:
Good spot. I just checked the source code for the method [here](https://github.com/python/cpython/blob/3.7/Lib/csv.py#L384) and it doesn't error as I thought, it returns a Boolean. So in my next commit I will add an `assert` statement before all `has_header()` calls.